overlord/servicestate/servicemgr.go

```diff
@@ -42,8 +56,128 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
 	return m
 }
 
+func MockEnsuredSnapServices(mgr *ServiceManager, ensured bool) (restore func()) {
+	osutil.MustBeTestBinary("ensured snap services can only be mocked from tests")
+	old := mgr.ensuredSnapSvcs
+	mgr.ensuredSnapSvcs = ensured
+	return func() {
+		mgr.ensuredSnapSvcs = old
+	}
+}
+
+func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
+	m.state.Lock()
+	defer m.state.Unlock()
+	if m.ensuredSnapSvcs {
+		return nil
+	}
+
+	// only run after we are seeded
+	var seeded bool
+	err = m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	// we are seeded, now we need to find all snap services and re-generate
+	// services as necessary
+
+	// ensure all snap services are updated
+	allStates, err := snapstate.All(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// if we have no snaps we can exit early
+	if len(allStates) == 0 {
+		m.ensuredSnapSvcs = true
+		return nil
+	}
+
+	snapsMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+
+		// don't use EnsureSnapServices with the snapd snap
+		if info.Type() == snap.TypeSnapd {
+			continue
+		}
+
+		snapSvcOpts, err := SnapServiceOptions(m.state, info.InstanceName())
+		if err != nil {
+			return err
+		}
+		snapsMap[info] = snapSvcOpts
+	}
+
+	// setup ensure options
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := snapstate.DeviceCtx(m.state, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	rewrittenServices := make(map[*snap.Info][]*snap.AppInfo)
+	serviceKillingMightHaveOccurred := false
+	observeChange := func(app *snap.AppInfo, unitType, name string, old, new string) {
+		if unitType == "service" {
+			rewrittenServices[app.Snap] = append(rewrittenServices[app.Snap], app)
+			if !serviceKillingMightHaveOccurred {
+				if strings.Contains(old, "\nRequires=usr-lib-snapd.mount\n") {
+					serviceKillingMightHaveOccurred = true
+				}
+			}
+		}
+	}
+
+	err = wrappers.EnsureSnapServices(snapsMap, ensureOpts, observeChange, progress.Null)
```

Comment:
Does it make sense to not rewrite antyhing if on core (pre-18) or on classic?

Reply:
I think we want the rewriting anyways, the rewriting will fix any existing devices that suffer from the vitality bug that I fixed in this PR: https://github.com/snapcore/snapd/pull/10163 and it will enable us to fix bugs like the docker CVE without lots of extra work and we have always wanted something like this anyways so I think we do want to re-generate services on all devices